### PR TITLE
better error message when api is loading

### PR DESCRIPTION
**Pull request by avahowell**
## before:
before
## after:
after

This should give users and API developers more helpful feedback when Sia is loading modules.

### CODE REVIEW

File: cmd/siad/server.go [hunk 1]
```diff
@@ -366,6 +368,21 @@ func (srv *Server) daemonHandler(password string) http.Handler {
...
+	api.WriteError(w, api.Error{"siad is not ready. please wait for siad to finish loading."}, http.StatusServiceUnavailable)
+}
+
+// AttachAPI attaches an initialized Sia API to the server.
+func (srv *Server) AttachAPI(a *api.API) {
+	mux := http.NewServeMux()
+	mux.Handle("/daemon/", api.RequireUserAgent(srv.daemonHandler(srv.requiredPassword), srv.requiredUserAgent))
+	mux.Handle("/", a)
+	srv.mux = mux
+	srv.httpServer.Handler = mux
```
Comment (lukechampine) [inline cmd/siad/server.go]: isn't this racey?

Comment (avahowell) [reply cmd/siad/server.go]: It may be. I'll write a test for this behavior that uses the race detector

**Comment by avahowell** [comment]
Latest commits fix the data race and restructure the code slightly so the daemon `Server` is responsible for loading and closing the modules.

File: cmd/siad/server.go [hunk 2]
```diff
@@ -422,9 +578,19 @@ func (srv *Server) Serve() error {
...
 	// Close the listener, which will cause Server.Serve() to return.
 	if err := srv.listener.Close(); err != nil {
-		return err
+		errs = append(errs, err)
 	}
-	return nil
+	// Close all of the modules
+	for i := len(srv.moduleClosers) - 1; i >= 0; i-- {
+		m := srv.moduleClosers[i]
+		fmt.Printf("Closing %v...\n", reflect.TypeOf(m))
```
Comment (lukechampine) [inline cmd/siad/server.go]: does this actually print the right thing?

Comment (avahowell) [reply cmd/siad/server.go]: Example output:

```
Caught stop signal, quitting...
Closing *renter.Renter...
Closing *host.Host...
Closing *wallet.Wallet...
Closing *transactionpool.TransactionPool...
Closing *consensus.ConsensusSet...
Closing *gateway.Gateway...
Shutdown complete.
```

Comment (lukechampine) [reply cmd/siad/server.go]: It's a cute trick but not good enough. Let's use a separate slice for the names, or define a type containing both the name and io.Closer.

Comment (avahowell) [reply cmd/siad/server.go]: ACK, I added a new `moduleCloser` type which prints the name in its `Close` method.

File: cmd/siad/server.go [hunk 3]
```diff
@@ -145,6 +151,12 @@ func (rs byVersion) Less(i, j int) bool {
 	return build.VersionCmp(rs[i].version(), rs[j].version()) >= 0
 }
 
+// Close implements the Close method for moduleCloser.
+func (mc *moduleCloser) Close() error {
+	fmt.Printf("Closing %v...\n", mc.name)
+	return mc.closer.Close()
+}
```
Comment (lukechampine) [inline cmd/siad/server.go]: hmm. Having in the `Printf` inside the `Close` feels wrong. I'd keep it outside.

Comment (lukechampine) [reply cmd/siad/server.go]: specifically, I'd embed the `io.Closer` in `moduleCloser`, so you don't need to define a new method at all

Comment (avahowell) [reply cmd/siad/server.go]: Ah, gotcha. I'll push another commit

**Pull request merged**